src/common/utility/clone-decorator.ts

```diff
@@ -0,0 +1,61 @@
+import { cloneDeep, memoize } from 'lodash';
+
+export default function cloneDecorator<T extends Method>(target: object, key: string, descriptor: TypedPropertyDescriptor<T>): TypedPropertyDescriptor<T>;
+export default function cloneDecorator<T extends Constructor<object>>(target: T): T;
+export default function cloneDecorator(target: any, key?: any, descriptor?: any): any {
+    if (!key || !descriptor) {
+        return cloneClassDecorator(target);
+    }
+
+    return cloneMethodDecorator(target, key, descriptor);
+}
+
+export function cloneClassDecorator<T extends Constructor<object>>(target: T): T {
+    const decoratedTarget = class extends target {};
+
+    Object.getOwnPropertyNames(target.prototype)
+        .forEach(key => {
+            const descriptor = Object.getOwnPropertyDescriptor(target.prototype, key);
+
+            if (!descriptor || key === 'constructor') {
+                return;
+            }
+
+            Object.defineProperty(
+                decoratedTarget.prototype,
+                key,
+                cloneMethodDecorator(target.prototype, key, descriptor)
+                );
+            });
+
+    return decoratedTarget;
+}
+
+export function cloneMethodDecorator<T extends Method>(target: object, key: string, descriptor: TypedPropertyDescriptor<T>): TypedPropertyDescriptor<T> {
+    if (typeof descriptor.value !== 'function') {
+        return descriptor;
+    }
+
+    let method = descriptor.value;
+    const memoizedCloneDeep = memoize(cloneDeep);
```

Comment:
We'll need to pass in a cache key resolver here because, by default, Lodash's memoize only uses the first parameter as its cache key. For example, if a memoized function accepts two parameters, and you call it twice (the first argument is the same but the second argument is different), you will get the cached result for the second call. That's not what you want as the computed result might be different.

We have a `CacheKeyResolver` already so maybe you can just use that?

Reply:
But `cloneDeep` only takes one param.

<img width="242" alt="image" src="https://user-images.githubusercontent.com/4542735/61192435-9943c880-a6f7-11e9-8cc0-430af5300547.png">

Reply:
Ah yep, sorry I misread. I thought you're memoizing the decorated method. Ok, yes you're right, what you have should be fine and you don't need to pass in a cache key resolver.

PS: What you might want to consider is to replace the default `memoize.Cache` with a `WeakMap`? That would help with the garbage collection of objects. Have to check with the browser support though (it seems we're okay?).